TaskStackBuilder not created

Currently I'm trying to use the TaskStackBuilder method, but i'm getting a class cast exception:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{co.and.andco.development/co.and.andco.SplashActivity}: java.lang.ClassCastException: android.app.TaskStackBuilder cannot be cast to android.content.Intent
                                                                              at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2665)
                                                                              at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726)
                                                                              at android.app.ActivityThread.-wrap12(ActivityThread.java)
                                                                              at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
                                                                              at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                              at android.os.Looper.loop(Looper.java:154)
                                                                              at android.app.ActivityThread.main(ActivityThread.java:6119)
                                                                              at java.lang.reflect.Method.invoke(Native Method)
                                                                              at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
                                                                              at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
                                                                           Caused by: java.lang.ClassCastException: android.app.TaskStackBuilder cannot be cast to android.content.Intent
                                                                              at co.and.andco.DeepLinkDelegate.dispatchFrom(DeepLinkDelegate.java:104)
                                                                              at co.and.andco.DeepLinkDelegate.dispatchFrom(DeepLinkDelegate.java:50)
                                                                              at co.and.andco.SplashActivity.onStart(SplashActivity.java:43)
                                                                              at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1248)
                                                                              at android.app.Activity.performStart(Activity.java:6696)
                                                                              at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2628)
                                                                              at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726) 
                                                                              at android.app.ActivityThread.-wrap12(ActivityThread.java) 
                                                                              at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477) 
                                                                              at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                              at android.os.Looper.loop(Looper.java:154) 
                                                                              at android.app.ActivityThread.main(ActivityThread.java:6119) 
                                                                              at java.lang.reflect.Method.invoke(Native Method) 
                                                                              at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886) 
                                                                              at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```


Digging into the crash:

```
  method = c.getMethod(entry.getMethod(), Context.class);
            if (method.getReturnType().equals(TaskStackBuilder.class)) {
              taskStackBuilder = (TaskStackBuilder) method.invoke(c, activity);
              if (taskStackBuilder.getIntentCount() == 0) {
                return createResultAndNotify(activity, false, uri, "Could not deep link to method: " + entry.getMethod() + " intents length == 0" );
              }
              newIntent = taskStackBuilder.editIntentAt(taskStackBuilder.getIntentCount()-1);
            } else {
              newIntent = (Intent) method.invoke(c, activity);
            }
```

It seems that the TaskStackBuilder has to be from the support library. Is it possible to relax this check or add a different error message?

--- a/deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java
+++ b/deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java
@@ -60,6 +60,7 @@ import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.AnnotationValue;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.Modifier;
 import javax.lang.model.element.PackageElement;
 import javax.lang.model.element.TypeElement;
@@ -132,10 +133,8 @@ public class DeepLinkProcessor extends AbstractProcessor {
         error(customAnnotation, "Prefix property cannot be empty");
       }
       prefixes.put(customAnnotation, prefix);
-      for (Element customAnnotatedElement
-          : roundEnv.getElementsAnnotatedWith(MoreElements.asType(customAnnotation))) {
-        customAnnotatedElements.add(customAnnotatedElement);
-      }
+      customAnnotatedElements.addAll(
+          roundEnv.getElementsAnnotatedWith(MoreElements.asType(customAnnotation)));
     }
 
     Set<Element> elementsToProcess = new HashSet<>(customAnnotatedElements);
@@ -154,6 +153,14 @@ public class DeepLinkProcessor extends AbstractProcessor {
         if (!methodModifiers.contains(Modifier.STATIC)) {
           error(element, "Only static methods can be annotated with @%s",
               DEEP_LINK_CLASS.getSimpleName());
+        }
+        ExecutableElement executableElement = MoreElements.asExecutable(element);
+        TypeElement returnType = MoreTypes.asTypeElement(executableElement.getReturnType());
+        String qualifiedName = returnType.getQualifiedName().toString();
+        if (!qualifiedName.equals("android.content.Intent")
+            && !qualifiedName.equals("android.support.v4.app.TaskStackBuilder")) {
+          error(element, "Only `Intent` or `android.support.v4.app.TaskStackBuilder` are supported."
+                  + " Please double check your imports and try again.");
         }
       }
 
@@ -270,7 +277,7 @@ public class DeepLinkProcessor extends AbstractProcessor {
       DeepLinkAnnotatedElement element = elements.get(i);
       String type = "DeepLinkEntry.Type." + element.getAnnotationType().toString();
       ClassName activity = ClassName.get(element.getAnnotatedElement());
-      Object method = element.getMethod() == null ? null : element.getMethod();
+      Object method = element.getMethod();
       String uri = element.getUri();
       initializer.add("new DeepLinkEntry($S, $L, $T.class, $S)$L\n",
           uri, type, activity, method, (i < totalElements - 1) ? "," : "");
@@ -494,7 +501,7 @@ public class DeepLinkProcessor extends AbstractProcessor {
             + "link to method: \" + entry.getMethod() + \" intents length == 0\" )")
         .endControlFlow()
         .addStatement("newIntent = taskStackBuilder."
-            + "editIntentAt(taskStackBuilder.getIntentCount()-1)")
+            + "editIntentAt(taskStackBuilder.getIntentCount() - 1)")
         .nextControlFlow("else")
         .addStatement("newIntent = (Intent) method.invoke(c, activity, parameters)")
         .endControlFlow()

--- a/deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java
+++ b/deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java
@@ -225,7 +225,7 @@ public class DeepLinkProcessorTest {
             + "import com.airbnb.deeplinkdispatch.DeepLink; "
             + "public class SampleActivity {"
             + "  @DeepLink(\"airbnb://host/{arbitraryNumber}\")"
-            + "  public Intent intentFromNoStatic(Context context){"
+            + "  public Intent intentFromNoStatic(Context context) {"
             + "    return new Intent();"
             + "  }"
             + "}"
@@ -359,5 +359,29 @@ public class DeepLinkProcessorTest {
                     + "    return null;\n"
                     + "  }"
                     + "}"));
+  }
+
+  @Test public void testNonAppCompatTaskStackBuilderClassErrorMessage() {
+    JavaFileObject sampleActivity = JavaFileObjects
+        .forSourceString("SampleActivity", "package com.example;"
+            + "import com.airbnb.deeplinkdispatch.DeepLink; "
+            + "import android.content.Context;\n"
+            + "import android.app.TaskStackBuilder;\n"
+            + "import android.content.Intent;\n"
+            + "public class SampleActivity {"
+            + "  @DeepLink(\"airbnb://host/{arbitraryNumber}\")"
+            + "  public static TaskStackBuilder intentFromNoStatic(Context context) {"
+            + "    return TaskStackBuilder.create(context);"
+            + "  }"
+            + "}"
+        );
+
+    assertAbout(javaSource())
+        .that(sampleActivity)
+        .processedWith(new DeepLinkProcessor())
+        .failsToCompile()
+        .withErrorContaining(
+            "Only `Intent` or `android.support.v4.app.TaskStackBuilder` are supported."
+                + " Please double check your imports and try again.");
   }
 }

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-PROJECT_VERSION=4.0.0-SNAPSHOT
+PROJECT_VERSION=3.1.0-SNAPSHOT
 PROJECT_GROUP_ID=com.airbnb
 PROJECT_URL=https://github.com/airbnb/deeplinkdispatch
 PROJECT_DESCRIPTION=Library designed to handle deep linking in an Android application.

